src/auth.rs

```diff
@@ -20,92 +19,96 @@ use tower_http::auth::AuthorizeRequest;
 
 use crate::{extractors::AuthContext, user_store::UserStore, AuthUser};
 
-/// Layer that provides session-based authentication via [`AuthContext`].
-#[derive(Debug, Clone)]
-pub struct AuthLayer<User, Store, Role = ()> {
-    store: Store,
+#[derive(Clone)]
+struct AuthState<Store, User, Role = ()> {
     key: Key,
+    store: Store,
     _user_type: PhantomData<User>,
     _role_type: PhantomData<Role>,
 }
 
-impl<User, Store, Role> AuthLayer<User, Store, Role>
+/// Layer that provides session-based authentication via [`AuthContext`].
+#[derive(Clone)]
+pub struct AuthLayer<Store, User, Role = ()> {
+    state: AuthState<Store, User, Role>,
+}
+
+impl<Store, User, Role> AuthLayer<Store, User, Role>
 where
+    Store: UserStore<Role, User = User>,
     User: AuthUser<Role>,
     Role: PartialOrd + PartialEq + Clone + Send + Sync + 'static,
-    Store: UserStore<Role, User = User>,
 {
     /// Creates a layer which will attach the [`AuthContext`] and `User` to
     /// requests via extensions.
     ///
     /// Note that the `secret` is used to derive a key for HMAC signing. For
     /// security reasons, this value **must** be securely generated.
     pub fn new(store: Store, secret: &[u8]) -> Self {
-        Self {
+        let state = AuthState {
             store,
             key: Key::new(HMAC_SHA512, secret),
             _user_type: PhantomData,
             _role_type: PhantomData,
-        }
+        };
+
+        Self { state }
     }
 }
 
-impl<User, Store, Inner, Role> Layer<Inner> for AuthLayer<User, Store, Role>
+impl<S, Store, User, Role> Layer<S> for AuthLayer<Store, User, Role>
 where
-    Role: PartialOrd + PartialOrd + PartialEq + Clone + Send + Sync + 'static,
-    User: AuthUser<Role>,
     Store: UserStore<Role>,
+    User: AuthUser<Role>,
+    Role: PartialOrd + PartialEq + Clone + Send + Sync + 'static,
 {
-    type Service = Auth<User, Store, Inner, Role>;
+    type Service = AuthService<S, Store, User, Role>;
 
-    fn layer(&self, inner: Inner) -> Self::Service {
-        Auth {
+    fn layer(&self, inner: S) -> Self::Service {
+        AuthService {
             inner,
-            layer: self.clone(),
+            state: self.state.clone(),
         }
     }
 }
 
-#[derive(Debug, Clone)]
-pub struct Auth<User, Store, Inner, Role = ()> {
-    inner: Inner,
-    layer: AuthLayer<User, Store, Role>,
+#[derive(Clone)]
+pub struct AuthService<S, Store, User, Role = ()> {
+    inner: S,
+    state: AuthState<Store, User, Role>,
 }
 
-impl<User, Store, Role, Inner, ReqBody, ResBody> Service<Request<ReqBody>>
-    for Auth<User, Store, Inner, Role>
+impl<S, ReqBody, ResBody, Store, User, Role> Service<Request<ReqBody>>
+    for AuthService<S, Store, User, Role>
 where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Clone + Send + 'static,
+    S::Future: Send + 'static,
+    ReqBody: Send + 'static,
+    ResBody: Default + Send + 'static,
```

Comment:
It would be nice to not require `Default` but I'm not sure how else we can return an error response.